src/state.rs

```diff
@@ -259,6 +283,188 @@ impl DemandProxyState {
+        }
+        let ip = self
+            .load_balance_for_hostname(dst_workload, src_workload, metrics)
+            .await?;
+        Ok(ip)
+    }
+
+    async fn load_balance_for_hostname(
+        &self,
+        workload: &Workload,
+        src_workload: &Workload,
+        metrics: Arc<proxy::Metrics>,
+    ) -> Result<IpAddr, Error> {
+        let mut state: DemandProxyState = self.clone();
+        let labels = OnDemandDnsLabels::new()
+            .with_destination(workload)
+            .with_source(src_workload);
+        let workload_uid = workload.uid.to_owned();
+        let hostname = workload.hostname.to_owned();
+        let rdns = match state.get_ips_for_hostname(&workload.hostname) {
+            Some(r) => {
+                metrics
+                    .as_ref()
+                    .on_demand_dns_cache_hits
+                    .get_or_create(&labels)
+                    .inc();
+                r
+            }
+            None => {
+                metrics
+                    .as_ref()
+                    .on_demand_dns_cache_misses
+                    .get_or_create(&labels)
+                    .inc();
+                match Self::resolve_on_demand_dns(self.to_owned(), workload).await {
+                    Ok(_) => {
+                        trace!(
+                            "system dns async resolution: task finished for {:?}",
+                            &workload_uid
+                        );
+                    }
+                    Err(e) => {
+                        warn!(
+                            "system dns async resolution: error for {:?} is: {:?}",
+                            &workload_uid, e
+                        );
+                        return Err(Error::NoResolvedAddresses(workload_uid));
+                    }
+                };
+                // try to get it again
+                let updated_rdns = state.get_ips_for_hostname(&hostname);
+                match updated_rdns {
+                    Some(rdns) => rdns,
+                    None => {
+                        return Err(Error::NoResolvedAddresses(workload_uid));
+                    }
+                }
+            }
+        };
+
+        // TODO: add more sophisticated routing logic, perhaps based on ipv4/ipv6 support underneath us.
+        // if/when we support that, this function may need to move to get access to the necessary metadata.
+        // Randomly pick an IP
+        // TODO: do this more efficiently, and not just randomly
+        let Some(ip) = rdns.ips.iter().choose(&mut rand::thread_rng()) else {
+            return Err(Error::EmptyResolvedAddresses(workload_uid));
+        };
+        Ok(*ip)
+    }
+
+    fn resolve_on_demand_dns(mut state: DemandProxyState, workload: &Workload) -> JoinHandle<()> {
+        let workload_uid = workload.uid.to_owned();
+        let hostname = workload.hostname.to_owned();
+        tokio::spawn(async move {
```

Comment:
why spawn?

Reply:
no good reason, was copied from other async code and i missed that during cleanup. made simpler in https://github.com/istio/ztunnel/pull/612/commits/454372bbe8873e4bfaf0a697d21394e41c476350